### PR TITLE
Feature lru cache wrapper

### LINKED ISSUE #9461: Add Custom Constructor to LRU Cache to handle Invalid Parameters
# 💎 Issue

### Description

Brought up by @kasey 
https://github.com/prysmaticlabs/prysm/pull/9422#r695045376

The LRU cache returns an error when failing parameter validation, however for all our caches we need the parameters 
to be valid and therefore panic in the event there is any failure. A better way would be to create a custom constructor around the LRU cache which panics if an error is returned so that we can avoid the `cache,err` pattern and simply return the cache.

### PR DESCRIPTION
**Pull request by marino39**
**What type of PR is this?**
 Other

**What does this PR do? Why is it needed?**

It adds a requested wrapper on github.com/hashicorp/golang-lru, so no error handling is required. If lru cache creation fails it panics. The lru creation can only fail if cache size provided is negative.

**Which issues(s) does this PR fix?**

Fixes #9461 

**Other notes for review**

The interface for golang-lru Cache has been introduced in order to simplify imports. The interface has been chosen because in case of the struct wrapper we would need to copy lru.Cache which may not be safe.

### CODE REVIEW

**Review by nisdas — CHANGES_REQUESTED** [review]
Hey thanks for opening the PR ! Do you mind reverting the interface change ? It would be preferable to keep the PR in scope for fixing the related issue rather than implementing other changes.

**Comment by marino39** [comment]
> Hey thanks for opening the PR ! Do you mind reverting the interface change ? It would be preferable to keep the PR in scope for fixing the related issue rather than implementing other changes.
> 
> Also in the linked issue, I should have been more clear. We just need a custom LRU constructor to panic when invalid parameters are provided. There will be no need to create a LRU wrapper struct for it.

Hey, I have added interface just to avoid importing the lru twice from different places, once when you want to define lru struct and the other one when you want to use constructor wrapper. However if you are fine with it then I can revert that and leave only constructor function wrappers.

File: beacon-chain/sync/pending_attestations_queue_test.go [hunk 1]
```diff
@@ -106,7 +106,6 @@ func TestProcessPendingAtts_HasBlockSaveUnAggregatedAtt(t *testing.T) {
 
 	require.NoError(t, beaconState.SetGenesisTime(uint64(time.Now().Unix())))
 
-	c, err := lru.New(10)
 	require.NoError(t, err)
```
Comment (marino39) [inline beacon-chain/sync/pending_attestations_queue_test.go:109]: Self Review: This err check should be removed.

**Comment by marino39** [comment]
@nisdas OK, this is done. Please have a look when you have a time.

File: shared/lruwrpr/lru_wrpr.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package lruwrpr
+
+import (
+	"fmt"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+// New creates an LRU of the given size.
+func New(size int) *lru.Cache {
```
Comment (terencechain) [inline shared/lruwrpr/lru_wrpr.go:9]: Please add tests for both happy and unhappy cases

Comment (marino39) [reply shared/lruwrpr/lru_wrpr.go:9]: Done

**Review by nisdas — COMMENTED** [review]
Thanks this looks better now, although there is no need to name the package `lruwrpr` we can simply call it `lru` here.

**Comment by marino39** [comment]
> Thanks this looks better now, although there is no need to name the package `lruwrpr` we can simply call it `lru` here.

Ok, however it will need to be labeled as such in imports as the package name will collide with hashicorps one.

```golang
import (
    lru "github.com/hashicorp/golang-lru"
    lruwrpr "github.com/prysmaticlabs/prysm/shared/lru"
)
``` 

To be honest I do not have any other sane idea to name it. So I will leave it as it is.

**Review by nisdas — APPROVED** [review]
Thanks for addressing my comments, it looks good to me

**Comment by marino39** [comment]
@nisdas I have fixed import issues reported by gocheck. Sorry, nothing really complained about that before.

**Comment by nisdas** [comment]
@marino39 We haven't enabled some of our CI runners here for your PR here which is why it wasn't triggered. They are running now.

**Comment by nisdas** [comment]
@marino39 Do you mind fixing the conflicts ? will approve again once that is done

**Comment by marino39** [comment]
> @marino39 Do you mind fixing the conflicts ? will approve again once that is done

It's done.

**Comment by marino39** [comment]
I accidently pushed changes with my old work username/email. I need to recreate PR to fix this.

**Pull request closed without merging**